File: py/0013-roman-to-integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        symbol_to_value = {
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000,
        }

        ans = 0
        prev = max(symbol_to_value.values()) + 1

        for c in s:
            curr = symbol_to_value[c]
            if curr > prev:
                ans -= 2 * prev
            ans += curr
            prev = curr

        return ans

File: cpp/0013-roman-to-integer.cpp
#include <string>

using uint = unsigned;

class Solution {
    public:
        [[nodiscard]] constexpr auto romanToInt(const std::string &s
        ) const noexcept -> uint {
            auto ans = 0U;
            auto prev = 1001U;

            for (const auto c : s) {
                const auto curr = [c]() {
                    switch (c) {
                        case 'I':
                            return 1U;
                        case 'V':
                            return 5U;
                        case 'X':
                            return 10U;
                        case 'L':
                            return 50U;
                        case 'C':
                            return 100U;
                        case 'D':
                            return 500U;
                        default /* 'M' */:
                            return 1000U;
                    }
                }();

                if (curr > prev) {
                    ans -= 2 * prev;
                }

                ans += curr;
                prev = curr;
            }

            return ans;
        }
};
